WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift

```diff
@@ -90,3 +107,82 @@ enum CardPresentPaymentAdaptedCollectOrderPaymentResult {
+                return
+            case .connectingFailed(let error, let retrySearch, let endSearch):
+                endSearch()
+            case .connectingFailedNonRetryable(let error, let endSearch):
+                endSearch()
+            case .connectingFailedUpdatePostalCode(let retrySearch, let endSearch):
+                endSearch()
+            case .connectingFailedChargeReader(let retrySearch, let endSearch):
+                endSearch()
+            case .connectingFailedUpdateAddress(let wcSettingsAdminURL, let retrySearch, let endSearch):
+                endSearch()
+            case .foundReader(let name, let connect, let continueSearch, let endSearch):
+                endSearch()
+            case .updateProgress(let requiredUpdate, let progress, let cancelUpdate):
+                // TODO: handle optional case if possible?
+                cancelUpdate?()
+            case .updateFailed(let tryAgain, let cancelUpdate):
+                cancelUpdate()
+            case .updateFailedNonRetryable(let cancelUpdate):
+                cancelUpdate()
+            case .updateFailedLowBattery(let batteryLevel, let cancelUpdate):
+                cancelUpdate()
+            case .preparingForPayment(cancelPayment: let cancelPayment):
+                cancelPayment()
+            case .tapSwipeOrInsertCard(inputMethods: let inputMethods, cancelPayment: let cancelPayment):
+                cancelPayment()
+            case .paymentSuccess(done: let done):
+                done()
+            case .paymentError(error: let error, tryAgain: let tryAgain, cancelPayment: let cancelPayment):
+                cancelPayment()
+            case .paymentErrorNonRetryable(error: let error, cancelPayment: let cancelPayment):
+                cancelPayment()
+            case .processing:
+                cancelPayment()
+            case .displayReaderMessage(message: let message):
+                cancelPayment()
+            /// An alert to notify the merchant that the transaction was cancelled using a button on the reader
+            case .cancelledOnReader:
+                cancelPayment()
+            /// Before reader connection
+            case .selectSearchType:
+                cancelReaderSearch()
+            /// Connection already completed, before attempting payment
+            case .validatingOrder:
+                return
+        }
+    }
+}
+
+private extension CardPresentPaymentCollectOrderPaymentUseCaseAdaptor {
+    func cancelReaderSearch() {
+        stores.dispatch(CardPresentPaymentAction.cancelCardReaderDiscovery() { [weak self] _ in
+            //            self?.returnSuccess(result: .canceled(cancellationSource))
+        })
+    }
+
+    func cancelReaderConnection() {
+        // TODO
+    }
+
+    func cancelPayment() {
+        stores.dispatch(CardPresentPaymentAction.cancelPayment() { [weak self] result in
+            // TODO: implement allowPassPresentation when Tap To Pay is supported
```

Comment:
Great point and clarification! Actually, this can't be easily implemented outside of the payment orchestrator because it has an internal token for this. I updated `cancelPayment` to call the one from the payment orchestrator in https://github.com/woocommerce/woocommerce-ios/pull/13100/commits/9c1c698edb56e0e74909ce4d42f57c95b7f7a671 to ensure it includes the wallet functionality recovery.